File: src/go/Sorting/quick_sort.go
package Sorting

import (
	"math/rand"
)

// O(n * log n)
func Quick(raw []int, ascDirection bool) []int {
	length := len(raw)

	rawCopy := make([]int, length)
	copy(rawCopy, raw)

	if length < 2 {
		return rawCopy
	}
	pivotK := rand.Intn(length)
	pivotV := rawCopy[pivotK]

	less, greater := []int{}, []int{}

	rawCopy = append(rawCopy[:pivotK], rawCopy[pivotK+1:]...)

	for i := range rawCopy {
		v := rawCopy[i]
		if v < pivotV {
			less = append(less, v)
		} else {
			greater = append(greater, v)
		}
	}
	if !ascDirection {
		less, greater = greater, less
	}

	return append(append(Quick(less, ascDirection), pivotV), Quick(greater, ascDirection)...)
}

File: src/Sorting/QuickSort.php
<?php

namespace GulDmitry\Algorithms\Sorting;

/**
 * O(n*log n)
 */
class QuickSort
{
    public function sort(array $raw, bool $ascDirection = true): array
    {
        if (count($raw) < 2) {
            return $raw;
        }

        $pivotKey = array_rand($raw);
        $pivotValue = $raw[$pivotKey];

        unset($raw[$pivotKey]);

        $less = array_filter($raw, function ($v) use ($pivotValue) {
            return $v < $pivotValue;
        });
        $greater = array_filter($raw, function ($v) use ($pivotValue) {
            return $v >= $pivotValue;
        });

        if (!$ascDirection) {
            list($less, $greater) = [$greater, $less];
        }
        return $this->sort($less, $ascDirection) + [$pivotKey => $pivotValue] + $this->sort($greater, $ascDirection);
    }
}
